# Task1: Data Cleaning and Preprocessing:

### **Objective:** `Clean and Prepare raw dataset (with null values , duplicates , inconsistent formats).`

### **Tools:** `python--(Pandas)`

# Dataset: CUSTOMER PERSONALITY ANALYSIS:

### - `I gathered this dataset from kaggle`
### -  kaggle Link: [link text](https://(https://www.kaggle.com/datasets/imakash3011/customer-personality-analysis))

### `Customer Personality analysis is the process of understanding a company ideal customers through detailed analysis of their demographics , lifestyle , and purchasing behaviour.This understanding allows business to tailor their products , services and marketing strategies to better meet the needs of specific customer groups.`

### `Rather than investing resources in promoting products to all customers  , bussiness can focus on specific customer segments thata are more likely to respond posirively.This targeted approach helps customer satisfaction , marketing efficiency , and overall business performance`

# Attributes

# People-Related Attributes
####  - id - `Customer's unique identifier`
####  - year_birth -  `Year of Birth`
####  - Education -  `Level of Education`
####  - Marital_status -  `Martial status`
####  - Income -  `yearly household income`
####  - Kidhome - `Number of children in the household`
####  - Teenhome -	`Number of teenagers in the household`
####  - Dt_Customer - `Date when the customer enrolled`
####  - Recency	Days - `since last purchase`
####  - Complain -  `1 if customer filed a complaint in 2 years, else 0`


# Product-Related Attributes
#### MntWines - `Amount spent on wine in last 2 years`
#### MntFruits -`Amount spent on fruits`
#### MntMeatProducts -`Amount spent on meat products`
#### MntFishProducts - `Amount spent on fish products`
#### MntSweetProducts	- `Amount spent on sweets`
#### MntGoldProds - `Amount spent on gold products`

# Promotion-Related Attributes
#### NumDealsPurchases - `Number of purchases made with a discount`
#### AcceptedCmp1-5 - 	`1 if offer accepted in campaign 1-5, else 0`
#### Response	- `1 if offer accepted in the last campaign, else 0`

# Place/Channel Attributes
#### NumWebPurchases	`Purchases made through the website`
#### NumCatalogPurchases - `Purchases made via catalog`
#### NumStorePurchases - `Purchases made directly in-store`
#### NumWebVisitsMonth - `Website visits in the last month`





# TARGET:
### 1. `Identifying and handling missing values`
### 2. `Removing duplicates values`
### 3. `Standaridize text values`
### 4. `Convert Dates to a Consistent Format`
### 5. `Clean Column Names`
### 6. `Check and Fix Data Types`

# Dataset

### -  `Loading and reading the Customer Personality Analysis dataset using pandas dataframe`

In [66]:
import pandas as pd #--> python library for data manipulation

# loading and reading the dataset
dataset = pd.read_csv("/content/drive/MyDrive/Internship/marketing_campaign.csv",sep="\t") #--> sep="\t" specifies that the separator in the CSV file is a tab character

print("First 5 rows:")
print(dataset.head()) #--> returns the first 5 rows of the dataframe
print()

print("Last Five rows:")
dataset.tail()  # -->returns the last 5 rows of the dataframe

First 5 rows:
     ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
0  04-09-2012       58       635  ...                  7             0   
1  08-03-2014       38        11  ...                  5             0   
2  21-08-2013       26       426  ...                  4             0   
3  10-02-2014       26        11  ...                  6             0   
4  19-01-2014       94       173  ...                  5             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
2235         1  13-06-2013       46       709  ...                  5   
2236         1  10-06-2014       56       406  ...                  7   
2237         0  25-01-2014       91       908  ...                  6   
2238         1  24-01-2014        8       428  ...                  3   
2239         1  15-10-2012       40        84  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
2235             0             0             0             0             0   
2236             0             0             0             1             0   
2237             0             1             0             0             0   
2238             0             0             0             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
2235         0              3         11         0  
2236         0              3         11         0  
2237         0              3         11         0  
2238         0              3         11         0  
2239         0              3         11         1  

[5 rows x 29 columns]

### - `Lets check the shape of the dataset`

In [67]:
dataset.shape #-->returns the shape of the dataset containing (rows , columns)

(2240, 29)

### `Here we have 2240 rows and 29 columns`

### - `lets check the what the columns we have`


In [68]:
print(f"Columns")
dataset.columns.tolist()

Columns


['ID',
 'Year_Birth',
 'Education',
 'Marital_Status',
 'Income',
 'Kidhome',
 'Teenhome',
 'Dt_Customer',
 'Recency',
 'MntWines',
 'MntFruits',
 'MntMeatProducts',
 'MntFishProducts',
 'MntSweetProducts',
 'MntGoldProds',
 'NumDealsPurchases',
 'NumWebPurchases',
 'NumCatalogPurchases',
 'NumStorePurchases',
 'NumWebVisitsMonth',
 'AcceptedCmp3',
 'AcceptedCmp4',
 'AcceptedCmp5',
 'AcceptedCmp1',
 'AcceptedCmp2',
 'Complain',
 'Z_CostContact',
 'Z_Revenue',
 'Response']

# Basic Information of the Dataset

In [69]:
dataset.info() # --> gives the basic information of the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

### - `By seeing basic info we have basically good data`
### - `But we have missing values in income only`
### - `data of customer account enrolled we have incorrect data type`
### - `We have some unwanted columns which are not useful that we are dropping`


In [70]:
# droping the unwanted columns:
dataset = dataset.drop(columns=["ID" , "Z_CostContact" , "Z_Revenue"])

### - `As we all know there is no use of id column so we are dropping that`
### - `and Z_costContact and Z_revenvue is having constant values so we are dropping`

# 1. `Identifying and handling missing values`

### - `Only income column has missing values`

In [71]:
# check how many are missing
print(dataset.isnull().sum())


# replacing null values with median (because median is good for skewed data)
dataset["Income"] = dataset["Income"].fillna(dataset["Income"].median())

# checking for missing values
dataset.isnull().sum()

Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Response                0
dtype: int64


Year_Birth             0
Education              0
Marital_Status         0
Income                 0
Kidhome                0
Teenhome               0
Dt_Customer            0
Recency                0
MntWines               0
MntFruits              0
MntMeatProducts        0
MntFishProducts        0
MntSweetProducts       0
MntGoldProds           0
NumDealsPurchases      0
NumWebPurchases        0
NumCatalogPurchases    0
NumStorePurchases      0
NumWebVisitsMonth      0
AcceptedCmp3           0
AcceptedCmp4           0
AcceptedCmp5           0
AcceptedCmp1           0
AcceptedCmp2           0
Complain               0
Response               0
dtype: int64

# 2. Removing Duplicates



### a. `check for duplicate first`

In [72]:
duplicted_count = dataset.duplicated().sum()
print(f"Number of duplicate rows: {duplicted_count}")

Number of duplicate rows: 182


### - `We have 182 duplicate rows now we are dropping them`

In [73]:
dataset = dataset.drop_duplicates()

In [74]:
# re-check
print(f"Checking the shape of dataset after dropping duplicates {dataset.shape}")

Checking the shape of dataset after dropping duplicates (2058, 26)


### -`Previously we have 2240 rows now we have 2058 rows`

# 3.`Standardize text values`

### a.`Check unique values before cleaning`

In [75]:
print(dataset['Education'].unique())
print(dataset['Marital_Status'].unique())

['Graduation' 'PhD' 'Master' 'Basic' '2n Cycle']
['Single' 'Together' 'Married' 'Divorced' 'Widow' 'Alone' 'Absurd' 'YOLO']


### `b.remove whitespace and apply title case`

In [76]:
dataset['Education'] = dataset['Education'].str.strip().str.title()
dataset['Marital_Status'] = dataset['Marital_Status'].str.strip().str.title()

### `c.Grouping similar values`

In [77]:
dataset['Marital_Status'] = dataset['Marital_Status'].replace({
    'Alone':'Single',
    'YOLO':'Single',
    'Absurd':'Single',
    'Widow':'Widowed',
    'Yolo':'Single'
})

### `d.Recheck for Unique values`

In [78]:
print(dataset['Education'].unique())
print(dataset['Marital_Status'].unique())

['Graduation' 'Phd' 'Master' 'Basic' '2N Cycle']
['Single' 'Together' 'Married' 'Divorced' 'Widowed']


# `4.Convert Dates to a Consistent Format`

### `a.Check the current type and sample values`

In [79]:
print(dataset['Dt_Customer'].dtype)
print(dataset['Dt_Customer'].head())

object
0    04-09-2012
1    08-03-2014
2    21-08-2013
3    10-02-2014
4    19-01-2014
Name: Dt_Customer, dtype: object


### - `It is an object type as we know date should be in dd-mm-year format`

In [80]:
dataset['Dt_Customer'] = pd.to_datetime(dataset['Dt_Customer'], format='%d-%m-%Y')

### `b.recheck`

In [81]:
print(dataset['Dt_Customer'].dtype)
print(dataset['Dt_Customer'].head())

datetime64[ns]
0   2012-09-04
1   2014-03-08
2   2013-08-21
3   2014-02-10
4   2014-01-19
Name: Dt_Customer, dtype: datetime64[ns]


# `5.Clean Column names`

In [82]:
dataset.columns = dataset.columns.str.strip()  #-->removing leading and trailing spaces
dataset.columns = dataset.columns.str.lower() #--> lower casing the columns
dataset.columns = dataset.columns.str.replace(' ' , '_')  #-->converting spaces with underscores
dataset.columns = dataset.columns.str.replace('[^a-z0-9_]' , '' , regex=True)  # remove special characters

In [83]:
print(dataset.columns)

Index(['year_birth', 'education', 'marital_status', 'income', 'kidhome',
       'teenhome', 'dt_customer', 'recency', 'mntwines', 'mntfruits',
       'mntmeatproducts', 'mntfishproducts', 'mntsweetproducts',
       'mntgoldprods', 'numdealspurchases', 'numwebpurchases',
       'numcatalogpurchases', 'numstorepurchases', 'numwebvisitsmonth',
       'acceptedcmp3', 'acceptedcmp4', 'acceptedcmp5', 'acceptedcmp1',
       'acceptedcmp2', 'complain', 'response'],
      dtype='object')


# `6.Check and fix datatypes`

### `a.Check current datatypes:`

In [84]:
print(dataset.dtypes)

year_birth                      int64
education                      object
marital_status                 object
income                        float64
kidhome                         int64
teenhome                        int64
dt_customer            datetime64[ns]
recency                         int64
mntwines                        int64
mntfruits                       int64
mntmeatproducts                 int64
mntfishproducts                 int64
mntsweetproducts                int64
mntgoldprods                    int64
numdealspurchases               int64
numwebpurchases                 int64
numcatalogpurchases             int64
numstorepurchases               int64
numwebvisitsmonth               int64
acceptedcmp3                    int64
acceptedcmp4                    int64
acceptedcmp5                    int64
acceptedcmp1                    int64
acceptedcmp2                    int64
complain                        int64
response                        int64
dtype: objec

### `b.Create and fix age column instead of birth year`

In [85]:
dataset['age'] = 2025 - dataset['year_birth']
dataset['age'] = dataset['age'].astype(int)

### `C.Income column as float`

In [86]:
dataset['income'] = dataset['income'].astype(float)

### `d.make sure binary column as int`

In [87]:
binary_columns = ['complain', 'response', 'acceptedcmp1', 'acceptedcmp2', 'acceptedcmp3',
                  'acceptedcmp4', 'acceptedcmp5']
dataset[binary_columns] = dataset[binary_columns].astype(int)

In [88]:
print(dataset.dtypes)

year_birth                      int64
education                      object
marital_status                 object
income                        float64
kidhome                         int64
teenhome                        int64
dt_customer            datetime64[ns]
recency                         int64
mntwines                        int64
mntfruits                       int64
mntmeatproducts                 int64
mntfishproducts                 int64
mntsweetproducts                int64
mntgoldprods                    int64
numdealspurchases               int64
numwebpurchases                 int64
numcatalogpurchases             int64
numstorepurchases               int64
numwebvisitsmonth               int64
acceptedcmp3                    int64
acceptedcmp4                    int64
acceptedcmp5                    int64
acceptedcmp1                    int64
acceptedcmp2                    int64
complain                        int64
response                        int64
age         

# `Interview question and answers`

### `1.What are missing values and how do you handle them?`
### `Ans: missing values are empty or null entries in a dataset where data is expected.`
## `handling methods include:`
### - `Removing rows / columns with too many missing values using dropna()`
### - `filling missing values  using fillna() with:`
###  - `Mean / median / mode`
###  - `Backward / forward fill`
###  - `Domain-specific constants(eg 0 or unknow)`




# 2. `How do you treat duplicate records?`

### `Duplicate records can lead to biased results or incorrect insights.`
### `To handle them:`

### - `Use .duplicated() to detect duplicates`
### - `Use .drop_duplicates() to remove them`

# `3.Difference between dropna() and fillna() in Pandas?`

### `- dropna()	Removes rows or columns with missing values`
### `- fillna()	Replaces missing values with a specified value`

# `4.What is outlier treatment and why is it important?`
### `Outliers are extreme values that differ significantly from others.`
**Importance:**

### - `They can skew statistics like mean, standard deviation`

### - `May impact model accuracy`

**Treatment Methods:**

### `Remove them using IQR or Z-score method`



# 5.`Explain the process of standardizing data.`
### `Standardization means transforming data to a standard scale (usually mean = 0 and std = 1).`

**Why?**

### - `Helps improve model performance, especially distance-based models like KMeans or KNN.`

# `6.How do you handle inconsistent data formats (e.g., date/time)?`

### - `Use pd.to_datetime() to convert strings to datetime`

### - `Standardize formats (e.g., dd-mm-yyyy)`

### - `Remove or correct invalid date entries`

### - `Use .astype() to convert columns to correct types`

# `7.What are common data cleaning challenges?`

### - `Missing or inconsistent data`

### - `Duplicate entries`

### - `Incorrect data types`

### - `Inconsistent formats (e.g., dates, currency)`

### - `Outliers or noise`

### - `Typos in categorical variables`

### - `Unstructured or mixed data in a single column`



# `8.How can you check data quality?`
### - `Use .info() and .describe() to inspect data types and statistics`

**Check for:**

### - `Missing values`

### - `Duplicates`

### - `Inconsistent formats`

### - `Outliers`

### - `Visualizations: boxplots, histograms, heatmaps`

### - `Domain knowledge: e.g., income shouldn't be negative`